File: python/generator_x4l.py
import cv2
import os
import time

left_cam = cv2.VideoCapture(2, cv2.CAP_V4L2)
right_cam = cv2.VideoCapture(4, cv2.CAP_V4L2)

left_cam.set(cv2.CAP_PROP_BUFFERSIZE, 1)
right_cam.set(cv2.CAP_PROP_BUFFERSIZE, 1)

count = 1
while True:
    for _ in range(5):
        left_cam.grab()
        right_cam.grab()
    
    retL, frameL = left_cam.retrieve()
    retR, frameR = right_cam.retrieve()
    
    if not retL or not retR:
        continue
    
    cv2.imshow("Press SPACE to capture", cv2.hconcat([frameL, frameR]))
    
    key = cv2.waitKey(1)
    if key == 32:  # Пробел
        timestamp = time.time()
        cv2.imwrite(f"../data/image/chessboard_10_7_paper_st_1/left{count}.png", frameL)
        cv2.imwrite(f"../data/image/chessboard_10_7_paper_st_1/right{count}.png", frameR)
        print(f"Saved pair {count}")
        count += 1
    elif key == ord('q'):
        break

left_cam.release()
right_cam.release()
cv2.destroyAllWindows()

File: src/generator_x4l.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <sys/stat.h>

int main() {
    std::string save_dir = "../data/image/chessboard_10_7_paper_st_1";
    mkdir(save_dir.c_str(), 0777);

    cv::VideoCapture left_cam(2, cv::CAP_V4L2);
    cv::VideoCapture right_cam(4, cv::CAP_V4L2);

    if (!left_cam.isOpened() || !right_cam.isOpened()) {
        std::cerr << "Error: Cannot open cameras" << std::endl;
        return 1;
    }

    left_cam.set(cv::CAP_PROP_BUFFERSIZE, 1);
    right_cam.set(cv::CAP_PROP_BUFFERSIZE, 1);

    cv::Mat frameL, frameR, combined;
    int count = 1;

    while (true) {
        for (int i = 0; i < 5; i++) {
            left_cam.grab();
            right_cam.grab();
        }

        bool retL = left_cam.retrieve(frameL);
        bool retR = right_cam.retrieve(frameR);

        if (!retL || !retR) {
            continue;
        }

        cv::hconcat(frameL, frameR, combined);
        cv::imshow("Press SPACE to capture", combined);

        int key = cv::waitKey(1);
        if (key == 32) {
            cv::imwrite(save_dir + "/left" + std::to_string(count) + ".png", frameL);
            cv::imwrite(save_dir + "/right" + std::to_string(count) + ".png", frameR);
            std::cout << "Saved pair " << count << std::endl;
            count++;
        } else if (key == 'q') {
            break;
        }
    }

    left_cam.release();
    right_cam.release();
    cv::destroyAllWindows();

    return 0;
}